stl/inc/syncstream

```diff
@@ -0,0 +1,329 @@
+                _Result = false;
+            }
+
+            if (_Record_sync) {
+                if (_Wrapped->pubsync() == -1) {
+                    _Result = false;
+                }
+            }
+        }
+        _Record_sync = false;
+        streambuf_type::setp(
+            _Begin_seq_ptr, pointer_traits<pointer>::pointer_to(*streambuf_type::epptr())); // reset written data
+        return _Result;
+    }
+
+    _NODISCARD streambuf_type* get_wrapped() const noexcept {
+        return _Wrapped;
+    }
+
+    _NODISCARD allocator_type get_allocator() const noexcept {
+        return _Getal();
+    }
+
+    void set_emit_on_sync(bool _Val) noexcept {
+        _Emit_on_sync = _Val;
+    }
+
+protected:
+    int sync() override {
+        _Record_sync = true;
+
+        if (_Emit_on_sync) {
+            if (!emit()) {
+                return -1;
+            }
+        }
+        return 0;
+    }
+
+    int_type overflow(int_type _Current_elem) override {
+        if (_Wrapped == nullptr) {
+            return _Traits::eof();
+        }
+        const bool _Chk_eof = _Traits::eq_int_type(_Current_elem, _Traits::eof());
+        if (_Chk_eof) {
+            return _Traits::not_eof(_Current_elem);
+        }
+
+        const size_type _Buf_size       = _Get_buffer_size();
+        const size_type _Max_allocation = allocator_traits<_Alloc>::max_size(_Al);
+        if (_Buf_size == _Max_allocation) {
+            return _Traits::eof();
+        }
+
+        const size_type _New_capacity  = _Calculate_growth(_Buf_size, _Buf_size + 1, _Max_allocation);
+        const pointer _Old_ptr         = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        const size_type _Old_data_size = _Get_data_size();
+
+        const pointer _New_ptr = _Al.allocate(_New_capacity);
+        _Traits::copy(_New_ptr, _Old_ptr, _Old_data_size);
+
+        streambuf_type::setp(_New_ptr, _New_ptr + _Old_data_size, _New_ptr + _New_capacity);
+        streambuf_type::sputc(_Traits::to_char_type(_Current_elem));
+
+        return _Current_elem;
+    }
+
+private:
+    static constexpr size_type _Min_size = 32; // constant for minimum buffer size
+
+    void _Init() {
+        const pointer _New_ptr = _Al.allocate(_Min_size);
+        streambuf_type::setp(_New_ptr, _New_ptr + _Min_size);
+    }
+
+    void _Tidy() noexcept {
+        const size_type _Buf_size = _Get_buffer_size();
+        _Al.deallocate(pointer_traits<pointer>::pointer_to(*streambuf_type::pbase()), _Buf_size);
+
+        streambuf_type::setp(nullptr, nullptr, nullptr);
+        if (_Wrapped != nullptr) {
+            _Release_mutex_for_instance(_Wrapped);
+            _Wrapped = nullptr;
+        }
+    }
+
+    void _Assign_rv(basic_syncbuf&& _Right) {
+        if (this != _STD addressof(_Right)) {
+            _Tidy();
+            this->_Swap_except_al(_Right);
+        }
+    }
+
+    void _Move_assign(basic_syncbuf&& _Right, _Equal_allocators) noexcept {
+        _Tidy();

```

Comment:
I will add the check to `operator=` after `_Emit` call. It seems like the better option since `_Move_assign` could do this check twice if `_No_propagate_allocators` and `_Al == _Right._Al` are both true.